text/0000-from-lossy.md

```diff
@@ -0,0 +1,321 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Type conversions can be handled by the following traits:
+
+- `From` for infallible, exact conversions (e.g. widening, `u8` → `u16`)
+- `TryFrom` for fallible, exact conversions (e.g. narrowing, `u16` → `u8`, and signed, `i16` → `u16`)
+- `FromLossy` for infallible, lossy conversions (mostly concerning floating-point types, e.g. `u32` → `f32`)
+- `TryFromLossy` for fallible, inexact conversions (e.g. `f32` → `u32`)
+- `TruncateFrom` for truncations (e.g. `u16` → `u8` which drops high bits)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### `From` trait
+
+Tweak the documentation to clarify this.
+
+Remove several implementations on SIMD types (currently including many fallible
+conversions like `f32x4` → `i8x4`; nightly only).
+
+### `FromLossy` trait
+
+Add `std::convert::FromLossy`:
+
+```rust
+/// A trait for conversions which are potentially inexact.
+/// 
+/// It is required that conversions do not fail, and required that results are
+/// either approximately equivalent or are an appropriate special value.
+/// For example, it is reasonable for an implementation converting `f64` to
+/// `f32` to lose precision and to return positive or negative infinity for
+/// out-of-range results, as well as to drop payloads from NaN values.
+pub trait FromLossy {
+    fn from_lossy(x: T) -> Self;
+}
+```
+
+Add implementations to `f64` from all of:
+
+- `u64, i64, u128, i128`
+
+and to `f32` from all of:
+
+- `f64, u32, i32, u64, i64, u128, i128`
+
+(Note: other integer → float conversions are
+already handled by `From`. There is a question below about trait overlap.)
+
+Where conversions are not exact, they should be equivalent to a conversion
+which first forces some number of low bits of the integer representation to
+zero, and then does an exact conversion to floating-point.
+
+### `TryFromLossy` trait
+
+Add `std::convert::TryFromLossy`:
+
+```rust
+/// A trait for conversions which may fail and may be inexact.
+/// 
+/// Implementations should fail when the output type has no suitable
+/// corresponding or general-purpose value, and return an approximately
+/// equivalent value when possible.
+pub trait TryFromLossy {
+    type Error;
+    fn try_from_lossy(x: T) -> Result<Self, Self::Error>;
+}
+```
+
+Add implementations from all of:
+
+- `f32, f64`
+
+to all of:
+
+- `u8, u16, u32, u64, u128`
+- `i8, i16, i32, i64, i128`
+
+(Note: `f32` → `u128` is infallible (but still lossy) *if* the input is not
+negative, infinite or an NaN. So even though the output type has large enough
+range, this conversion trait is still applicable.)
+
+The implementations should fail on NaN, Inf, negative values (for unsigned
+integer result types), and values whose integer approximation is not
+representable. The integer approximation should be the value rounded towards
```

Comment:
Failing if the integer approximation doesn't fit the target type is... debatable. The proposed semantics for `as` are to saturate to `int_type::MAX` or `int_type::MIN` respectively, and that *is* in some ways an "approximately equivalent value".

BTW the same point applies to infinities (i.e., it's reasonable to define `f32::INFINITY -> u8` to result in 255), but I can understand being more uneasy with that.

Reply:
No, infinities and out-of-range values fall into the same category here: not representable. IMO 255 is a very poor approximation of 1000f32 and an even worse approximation to 1e300f64.

Certainly I'm being opinionated here but I think if we have a fallible conversion available then we should fail on out-of-range values. (But if not, then `TryFromLossy` also has no reason to exist.)

Reply:
I have sympathy with that argument, but at the same time I'd like to reduce divergence from the semantics of `as` (to avoid people sticking to `as` because they prefer its behavior), and currently that seems more likely to be saturation than panicking.

Reply:
Also note that `as` *must* return some defined value simply in order to avoid undefined behaviour. We have no need to model this conversion trait on the limitations of `as`.

Reply:
`as` doesn't have to return a value, it can panic (there are some reasons to not want that, but it avoids UB and is reasonable in isolation). That would be the moral equivalent of this trait returning `Err`.